File: khs/dijkstra/미확인도착지.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <algorithm>

using namespace std;
#define INF 10000000
#define MAX_N 2001

typedef pair <int, int> P;
typedef vector <P> V;
// typedef tuple <int, int , int> T;
int g, h;
struct T{
    int first, second, third;

};

int dis[MAX_N];
V graph[MAX_N];
P check[MAX_N];

struct Compare{
    bool operator()(const T &a, const T &b){
        if (a.first != b.first)
            return a.first > b.first;
        else if(a.second != b.second)
            return a.second > b.second;
        return a.third > b.third;
    }
};

void dijkstra(int start){
    priority_queue<T, vector<T>, Compare> pq;
    T t = {0, start, 0};
    pq.push(t);
    

    while(!pq.empty()){
        auto t = pq.top();
        int d, now, p;
        d = t.first;
        now = t.second;
        p = t.third;
        pq.pop();

        if (dis[now] < d)
            continue;

        for (auto c = graph[now].begin(); c < graph[now].end(); c++){
            int _d = d + c->second;
            int temp_p = p;

            if ((now == g && c->first == h) or (now == h && c->first == g))
                temp_p = 1;

            if (_d <= dis[c->first]){
                if (dis[c->first] == _d && check[c->first].first == 1)
                    continue;
                check[c->first] = {temp_p, _d};
                dis[c->first] = _d;
                T pq_push = {_d, c->first, temp_p};
                pq.push(pq_push);

            }
        }

    }

}


int main(){
    int T, n, m, t, s, a, b, d, x;
    cin >> T;

    while (T--){
        fill(dis, dis + MAX_N, 0);
        fill(graph, graph + MAX_N, V()); // 벡터의 경우 따로 초기화
        fill(check, check + MAX_N, P(0, 0));
        cin >> n >> m >> t;
        for  (int i = 0; i < n + 1; i++){
            dis[i] = INF;
        }
        cin >> s >> g >> h;
        for  (int i = 0; i < n + 1; i++){
            check[i] = {0,0};
        }
        dis[s] = 0;

        while (m--){
            cin >> a >> b >> d;
            graph[a].push_back({b,d});
            graph[b].push_back({a,d});
        }
        dijkstra(s);

        // for (int i = 0; i < n + 1; i++){
        //     for (auto j = graph[i].begin(); j < graph[i].end(); j++){
        //         cout << '(' << j->first << " " << j->second << ')';
        //     }
        //     cout << endl;
        // }

        // for (int i = 0; i < n + 1; i++){
        //     cout << dis[i] << " ";
        // }
        // cout << endl;

        // for (int i = 0; i < n + 1; i++){
        //     cout << '(' << check[i].first << " " << check[i].second  << ')';
        // }
        // cout << endl;






        priority_queue<int, vector<int>, greater<int>> q;
        while (t--){
            cin >> x;
            if (check[x].first == 1)
                q.push(x);
        }
        while (!q.empty()){
            cout << q.top() << " ";
            q.pop();
        }
        cout << "\n";
        
    }
}

File: khs/dijkstra/미확인도착지.py
import heapq

INF = int(10e9)

T = int(input())



for _ in range(T):
    n, m, t = map(int, input().split())
    
    dis = [INF] * (n + 1)
    graph = [[] for _ in range(n + 1)]
    s, g, h = map(int, input().split())
    check = [(0,0)] * (n + 1) # p, dis
    dis[s] = 0
    
    for _ in range(m):
        a, b, d = map(int, input().split())
        graph[a].append((b,d))
        graph[b].append((a,d))
    
    def dijkstra(start):
        q = []
        heapq.heappush(q, (0, start, 0))
        
        while q:
            # print(q)
            d, now, p = heapq.heappop(q)
            if dis[now] < d:
                continue
            
            for c in graph[now]:
                _d = d + c[1]
                temp_p = p
                if (now, c[0]) == (g, h) or (now, c[0]) == (h, g):
                    temp_p = 1
                if _d <= dis[c[0]]:
                    if dis[c[0]] == _d:
                        if check[c[0]][0] == 1:
                            # print(d, now, p)
                            continue
                    check[c[0]] = (temp_p, _d)

                    dis[c[0]] = _d
                    heapq.heappush(q, (_d ,c[0], temp_p))
                    
                    
    
    dijkstra(s)
    print(graph)
    print(dis)
    print(check)
 
    q = [] 
    for _ in range(t):
        x = int(input())
        if check[x][0] == 1:
            heapq.heappush(q, x)
            
    while q:
        print(heapq.heappop(q) , end= ' ')